source/_posts/2018/04/getting-started-with-vuex.md

```diff
@@ -0,0 +1,110 @@
+---
+title: Getting started with Vuex
+date: 2018-04-24
+categories: javascript
+tags: 
+  - vuejs
+  - frontend
+  - vuex
+---
+Tháng này mình muốn giới thiệu cho mọi người một library của Vue có khá nhiều lợi ích: `VUEX`
+### Vuex là gì ?
+
+Vuex được xem như là một thư viện dùng để quản lý `state` cho các ứng dụng của Vue.
+
+`State được hiểu như là trạng thái của của ứng dụng`. Ví dụ như app sẽ có trạng thái là đã login hay chưa.
+
+### Tại sao lại cần phải dùng Vuex ?
+Trước  tiên về `Vue`. Vue chia nhỏ `state` và quản lý trong từng component, điều này mang đến nhiều lợi ích như: dễ dùng lại code, giảm sự phức tạp của ứng dụng, việc test trở nên dễ dàng hơn, ... Nhưng một ngày đẹp trời có 3,4 ...n component dùng chung một state thì phải làm sao ?
+
+ Cách dễ dàng nhất là mỗi component đều sử dụng event send state ra ngoài để các component khác sử dụng và tự trigger khi state này thay đổi. Ngoài ra thì có thể tạo một component cha làm nơi trung chuyển state. Hai cách này đều làm cho work flow trở nên phức tạp hơn nhiều và khó có thể kiểm soát (điển hình nhất là khó khăn trong việc debug).
+
+Vuex ra đời để giải quyết những điều này :))
+
+### Kiến trúc của Vuex 
+
+![](/images/vuex/vuex.png)
+Với Vuex, state của ứng dụng được quản lý tập trung một chỗ được gọi là `store`, nó bao gồm 5 phần chính :
++ State 
++ Getters
++ Mutation
++ Actions
++ Modules
+
+#### State
+Cái tên nói lên tất cả :)) State là một object chứa toàn bộ state của ứng dụng
+
+```javascript
+const state = {
+  numbers: [1, 2, 3, 4, 5]
+}
+```
+#### Getters
+Là một tập các hàm dùng để filter state ra được output như mong muốn.
+
+```javascript
+const state = {
+  // không có tham số truyền vào 
+  getEvenNumbers: state => {
+    return state.numbers.filter(item => item%2 === 0)
+  }
+  // có tham số truyền vào 
+  hasNumber : state => num => {
+    return state.numbers.include(num)
+  }
+}
+```
+Để sử dụng trong component thì có thể gọi trực tiếp `this.$store.getters.getEvenNumbers` hoặc sử dụng `mapGetter`
+
+Chi tiết : [Getters](https://vuex.vuejs.org/en/getters.html)
+
+#### Mutations
+Đây là nơi duy nhất cho phép thay đổi state. Một điều cần lưu ý là `mutations` là `synchronous`. Vì sao phải cần điều này ? Nếu không có rule này thì trong trường hợp nhiều action cùng gọi 1 mutation làm sao ta kiểm soát được cái nào đã thay đổi state hay thứ tự thay đổi state như nào.
+
+```javascript
+const mutations = {
+  add(state, num){
+    return state.numbers.push(num)
+  }
+  remove(state, num) {
+    return state.numbers.include(num)
+  }
+}
+```
+
+### Actions
+Actions có thể nói là nơi thể hiện `bussiness logic`. Nó là nơi gọi API, save vào database, thực hiện các `commit` (gọi mutation) để thay đổi state, ... 
```

Comment:
Em thấy chỗ này không đúng.
Actions are similar to mutations, the differences being that:
- Instead of mutating the state, actions commit mutations.
- Actions can contain arbitrary asynchronous operations.